plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,251 @@
+package snmp_trap
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+
+	"github.com/soniah/gosnmp"
+)
+
+type handler func(*gosnmp.SnmpPacket, *net.UDPAddr)
+type execer func(string, ...string) ([]byte, error)
+
+type mibEntry struct {
+	mibName string
+	oidText string
+}
+
+type SnmpTrap struct {
+	ServiceAddress string `toml:"service_address"`
+
+	acc      telegraf.Accumulator
+	listener *gosnmp.TrapListener
+	timeFunc func() time.Time
+	errCh    chan error
+
+	makeHandlerWrapper func(handler) handler
+
+	Log telegraf.Logger `toml:"-"`
+
+	cacheLock sync.Mutex
+	cache     map[string]mibEntry
+
+	execCmd execer
+}
+
+var sampleConfig = `
+  ## Transport, local address, and port to listen on.  Transport must
+  ## be "udp://".  Omit local address to listen on all interfaces.
+  ##   example: "udp://127.0.0.1:1234"
+  # service_address = udp://:162
+`
+
+func (s *SnmpTrap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *SnmpTrap) Description() string {
+	return "Receive SNMP traps"
+}
+
+func (s *SnmpTrap) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func init() {
+	inputs.Add("snmp_trap", func() telegraf.Input {
+		return &SnmpTrap{
+			timeFunc:       time.Now,
+			ServiceAddress: "udp://:162",
+		}
+	})
+}
+
+func realExecCmd(arg0 string, args ...string) ([]byte, error) {
+	return exec.Command(arg0, args...).Output()
+}
+
+func (s *SnmpTrap) Init() error {
+	s.cache = map[string]mibEntry{}
+	s.execCmd = realExecCmd
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+
+	// wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	split := strings.SplitN(s.ServiceAddress, "://", 2)
```

Comment:
I'd have probably used `url.Parse`

Reply:
Thanks for the suggestion.  I'll leave it because it's about as much work to use url.Parse and then build a new string with the host and port.  If we ever support other transports we should switch to url.Parse